File: src/main/org/nlogo/agent/VertCylinder.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.agent;

import org.nlogo.api.AgentException;
import org.nlogo.api.AgentKindJ;

//world wraps along x-axis but not y-axis
final strictfp class VertCylinder
    extends Topology {
  VertCylinder(World world) {
    super(world);
  }


  //wrapping coordinates

  @Override
  double wrapX(double x) {
    return wrap(x, world.minPxcor() - 0.5, world.maxPxcor() + 0.5);
  }

  @Override
  double wrapY(double y)
      throws AgentException {
    double max = world.maxPycor() + 0.5;
    double min = world.minPycor() - 0.5;
    if (y >= max || y < min) {
      throw new AgentException("Cannot move turtle beyond the world's edge.");
    }
    return y;
  }

  @Override
  double distanceWrap(double dx, double dy, double x1, double y1, double x2, double y2) {
    double dx2 = x1 > x2 ? (x2 + world.worldWidth()) - x1 :
        (x2 - world.worldWidth()) - x1;
    dx = StrictMath.abs(dx2) < StrictMath.abs(dx) ? dx2 : dx;

    return world.rootsTable.gridRoot(dx * dx + dy * dy);
  }

  @Override
  double towardsWrap(double headingX, double headingY) {
    headingX = wrap(headingX, (-(double) world.worldWidth() / 2.0),
        (world.worldWidth() / 2.0));

    if (headingY == 0) {
      return headingX > 0 ? 90 : 270;
    }
    if (headingX == 0) {
      return headingY > 0 ? 0 : 180;
    }
    return (270 + StrictMath.toDegrees
        (StrictMath.PI + StrictMath.atan2(-headingY, headingX)))
        % 360;
  }

  @Override
  Patch getPatchAt(double xc, double yc)
      throws AgentException {

    if ((yc > world.maxPycor() + 0.5) ||
        (yc < world.minPycor() - 0.5)) {
      return null;
    }

    return world.getPatchAt(xc, yc);
  }

  @Override
  double shortestPathX(double x1, double x2) {
    double xprime;

    if (x1 > x2) {
      xprime = x2 + world.worldWidth();
    } else {
      xprime = x2 - world.worldWidth();
    }

    if (StrictMath.abs(x2 - x1) > StrictMath.abs(xprime - x1)) {
      x2 = xprime;
    }

    return x2;
  }

  @Override
  double shortestPathY(double y1, double y2) {
    return y2;
  }

  @Override
  void diffuse(double diffuseparam, int vn)
      throws AgentException, PatchException {
    int xx = world.worldWidth();
    int yy = world.worldHeight();
    int xx2 = xx * 2;
    int yy2 = yy * 2;
    double[][] scratch = world.getPatchScratch();
    double[][] scratch2 = new double[xx][yy];
    int x = 0, y = 0;
    int minx = world.minPxcor();
    int miny = world.minPycor();

    try {
      for (y = 0; y < yy; y++) {
        for (x = 0; x < xx; x++) {
          scratch[x][y] =
              ((Double) world.fastGetPatchAt(x + minx, y + miny)
                  .getPatchVariable(vn))
                  .doubleValue();
          scratch2[x][y] = 0;
        }
      }
    } catch (ClassCastException ex) {
      throw new PatchException(world.fastGetPatchAt
          ((int) wrapX(x), (int) wrapY(y)));
    }

    for (y = yy; y < yy2; y++) {
      for (x = xx; x < xx2; x++) {

        double diffuseVal = (scratch[x - xx][y - yy] / 8) * diffuseparam;

        if (y > yy && y < yy2 - 1) {
          scratch2[x - xx][y - yy] += scratch[x - xx][y - yy] - (8 * diffuseVal);
          scratch2[(x - 1) % xx][(y - 1) % yy] += diffuseVal;
          scratch2[(x - 1) % xx][y % yy] += diffuseVal;
          scratch2[(x - 1) % xx][(y + 1) % yy] += diffuseVal;
          scratch2[x % xx][(y + 1) % yy] += diffuseVal;
          scratch2[x % xx][(y - 1) % yy] += diffuseVal;
          scratch2[(x + 1) % xx][(y - 1) % yy] += diffuseVal;
          scratch2[(x + 1) % xx][y % yy] += diffuseVal;
          scratch2[(x + 1) % xx][(y + 1) % yy] += diffuseVal;
        } else if (y == yy) {
          scratch2[x - xx][y - yy] += scratch[x - xx][y - yy] - (5 * diffuseVal);
          scratch2[(x - 1) % xx][y % yy] += diffuseVal;
          scratch2[(x - 1) % xx][(y + 1) % yy] += diffuseVal;
          scratch2[x % xx][(y + 1) % yy] += diffuseVal;
          scratch2[(x + 1) % xx][y % yy] += diffuseVal;
          scratch2[(x + 1) % xx][(y + 1) % yy] += diffuseVal;
        } else {
          scratch2[x - xx][y - yy] += scratch[x - xx][y - yy] - (5 * diffuseVal);
          scratch2[(x - 1) % xx][y % yy] += diffuseVal;
          scratch2[(x - 1) % xx][(y - 1) % yy] += diffuseVal;
          scratch2[x % xx][(y - 1) % yy] += diffuseVal;
          scratch2[(x + 1) % xx][y % yy] += diffuseVal;
          scratch2[(x + 1) % xx][(y - 1) % yy] += diffuseVal;
        }
      }
    }

    for (y = 0; y < yy; y++) {
      for (x = 0; x < xx; x++) {
        if (scratch2[x][y] != scratch[x][y]) {
          world.getPatchAtWrap(x + minx, y + miny)
              .setPatchVariable(vn, Double.valueOf(scratch2[x][y]));
        }
      }
    }
  }

  @Override
  void diffuse4(double diffuseparam, int vn)
      throws AgentException, PatchException {
    int xx = world.worldWidth();
    int yy = world.worldHeight();
    int xx2 = xx * 2;
    int yy2 = yy * 2;
    double[][] scratch = world.getPatchScratch();
    double[][] scratch2 = new double[xx][yy];
    int x = 0, y = 0;
    int minx = world.minPxcor();
    int miny = world.minPycor();

    try {
      for (y = 0; y < yy; y++) {
        for (x = 0; x < xx; x++) {
          scratch[x][y] =
              ((Double) world.fastGetPatchAt(x + minx, y + miny)
                  .getPatchVariable(vn))
                  .doubleValue();
          scratch2[x][y] = 0;
        }
      }
    } catch (ClassCastException ex) {
      throw new PatchException(world.fastGetPatchAt
          ((int) wrapX(x), (int) wrapY(y)));
    }

    for (y = yy; y < yy2; y++) {
      for (x = xx; x < xx2; x++) {

        double diffuseVal = (scratch[x - xx][y - yy] / 4) * diffuseparam;

        if (y > yy && y < yy2 - 1) {
          scratch2[x - xx][y - yy] += scratch[x - xx][y - yy] - (4 * diffuseVal);
          scratch2[(x - 1) % xx][y % yy] += diffuseVal;
          scratch2[x % xx][(y + 1) % yy] += diffuseVal;
          scratch2[x % xx][(y - 1) % yy] += diffuseVal;
          scratch2[(x + 1) % xx][y % yy] += diffuseVal;
        } else if (y == yy) {
          scratch2[x - xx][y - yy] += scratch[x - xx][y - yy] - (3 * diffuseVal);
          scratch2[(x - 1) % xx][y % yy] += diffuseVal;
          scratch2[x % xx][(y + 1) % yy] += diffuseVal;
          scratch2[(x + 1) % xx][y % yy] += diffuseVal;
        } else {
          scratch2[x - xx][y - yy] += scratch[x - xx][y - yy] - (3 * diffuseVal);
          scratch2[(x - 1) % xx][y % yy] += diffuseVal;
          scratch2[x % xx][(y - 1) % yy] += diffuseVal;
          scratch2[(x + 1) % xx][y % yy] += diffuseVal;
        }
      }
    }

    for (y = 0; y < yy; y++) {
      for (x = 0; x < xx; x++) {
        if (scratch2[x][y] != scratch[x][y]) {
          world.getPatchAtWrap(x + minx, y + miny)
              .setPatchVariable(vn, Double.valueOf(scratch2[x][y]));
        }
      }
    }

  }

  @Override
  AgentSet getNeighbors(Patch source) {
    if (source.pycor == world.maxPycor()) {
      if (source.pycor == world.minPycor()) {
        if (source.pxcor == world.maxPxcor() && source.pxcor == world.minPxcor()) {
          return new ArrayAgentSet(AgentKindJ.Patch(),
              new Agent[]{},
              world);
        } else {
          return new ArrayAgentSet(AgentKindJ.Patch(),
              new Agent[]{getPatchEast(source),
                  getPatchWest(source)},
              world);
        }
      } else {
        if (source.pxcor == world.maxPxcor() && source.pxcor == world.minPxcor()) {
          return new ArrayAgentSet(AgentKindJ.Patch(),
              new Agent[]{getPatchSouth(source)},
              world);
        } else {
          return new ArrayAgentSet(AgentKindJ.Patch(),
              new Agent[]{getPatchEast(source), getPatchSouth(source),
                  getPatchWest(source), getPatchSouthEast(source),
                  getPatchSouthWest(source)},
              world);
        }
      }
    } else if (source.pycor == world.minPycor()) {
      if (source.pxcor == world.maxPxcor() && source.pxcor == world.minPxcor()) {
        return new ArrayAgentSet(AgentKindJ.Patch(),
            new Agent[]{getPatchNorth(source)},
            world);
      } else {
        return new ArrayAgentSet(AgentKindJ.Patch(),
            new Agent[]{getPatchNorth(source), getPatchEast(source),
                getPatchWest(source), getPatchNorthEast(source),
                getPatchNorthWest(source)},
            world);
      }
    } else {
      if (source.pxcor == world.maxPxcor() && source.pxcor == world.minPxcor()) {
        return new ArrayAgentSet(AgentKindJ.Patch(),
            new Agent[]{getPatchNorth(source), getPatchSouth(source)},
            world);
      } else {
        return new ArrayAgentSet(AgentKindJ.Patch(),
            new Agent[]{getPatchNorth(source), getPatchEast(source),
                getPatchSouth(source), getPatchWest(source),
                getPatchNorthEast(source), getPatchSouthEast(source),
                getPatchSouthWest(source), getPatchNorthWest(source)},
            world);
      }
    }
  }

  @Override
  double observerY() {
    return 0.0;
  }

  @Override
  double followOffsetY() {
    return 0.0;
  }

  @Override
  AgentSet getNeighbors4(Patch source) {
    if (source.pycor == world.maxPycor()) {
      if (source.pycor == world.minPycor()) {
        if (source.pxcor == world.maxPxcor() && source.pxcor == world.minPxcor()) {
          return new ArrayAgentSet(AgentKindJ.Patch(),
              new Agent[]{},
              world);
        } else {
          return new ArrayAgentSet(AgentKindJ.Patch(),
              new Agent[]{getPatchEast(source),
                  getPatchWest(source)},
              world);
        }
      } else {
        if (source.pxcor == world.maxPxcor() && source.pxcor == world.minPxcor()) {
          return new ArrayAgentSet(AgentKindJ.Patch(),
              new Agent[]{getPatchSouth(source)},
              world);
        } else {
          return new ArrayAgentSet(AgentKindJ.Patch(),
              new Agent[]{getPatchEast(source),
                  getPatchSouth(source),
                  getPatchWest(source)},
              world);
        }
      }
    } else if (source.pycor == world.minPycor()) {
      if (source.pxcor == world.maxPxcor() && source.pxcor == world.minPxcor()) {
        return new ArrayAgentSet(AgentKindJ.Patch(),
            new Agent[]{getPatchNorth(source)},
            world);
      } else {
        return new ArrayAgentSet(AgentKindJ.Patch(),
            new Agent[]{getPatchNorth(source),
                getPatchEast(source),
                getPatchWest(source)},
            world);
      }
    } else {
      if (source.pxcor == world.maxPxcor() && source.pxcor == world.minPxcor()) {
        return new ArrayAgentSet(AgentKindJ.Patch(),
            new Agent[]{getPatchNorth(source),
                getPatchSouth(source)},
            world);
      } else {
        return new ArrayAgentSet(AgentKindJ.Patch(),
            new Agent[]{getPatchNorth(source), getPatchEast(source),
                getPatchSouth(source), getPatchWest(source)},
            world);
      }
    }
  }

  //get patch

  @Override
  Patch getPN(Patch source) {
    if (source.pycor == world.maxPycor()) {
      return null;
    }

    return getPatchNorth(source);
  }

  @Override
  Patch getPE(Patch source) {
    return getPatchEast(source);
  }

  @Override
  Patch getPS(Patch source) {
    if (source.pycor == world.minPycor()) {
      return null;
    }

    return getPatchSouth(source);
  }

  @Override
  Patch getPW(Patch source) {
    return getPatchWest(source);
  }

  @Override
  Patch getPNE(Patch source) {
    if (source.pycor == world.maxPycor()) {
      return null;
    }

    return getPatchNorthEast(source);
  }

  @Override
  Patch getPSE(Patch source) {
    if (source.pycor == world.minPycor()) {
      return null;
    }

    return getPatchSouthEast(source);
  }

  @Override
  Patch getPSW(Patch source) {
    if (source.pycor == world.minPycor()) {
      return null;
    }

    return getPatchSouthWest(source);
  }

  @Override
  Patch getPNW(Patch source) {
    if (source.pycor == world.maxPycor()) {
      return null;
    }

    return getPatchNorthWest(source);
  }
}

File: src/main/org/nlogo/agent/VertCylinder.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.agent

import org.nlogo.api.{ AgentException, AgentKind }

// imagine a cylinder standing on end.

@annotation.strictfp
class VertCylinder(_world: World)
extends Topology(_world, xWraps = true, yWraps = false) {

  override def wrapX(x: Double): Double =
    Topology.wrap(x, world.minPxcor - 0.5, world.maxPxcor + 0.5)

  @throws(classOf[AgentException])
  override def wrapY(y: Double): Double = {
    val max = world.maxPycor + 0.5
    val min = world.minPycor - 0.5
    if (y >= max || y < min)
      throw new AgentException("Cannot move turtle beyond the world's edge.")
    y
  }

  override def distanceWrap(dx: Double, dy: Double, x1: Double, y1: Double, x2: Double, y2: Double): Double = {
    val dx2 =
      if (x1 > x2)
        x2 + world.worldWidth - x1
      else
        x2 - world.worldWidth - x1
    val dxMin =
      if (StrictMath.abs(dx2) < StrictMath.abs(dx))
        dx2
      else
        dx
    world.rootsTable.gridRoot(dxMin * dxMin + dy * dy)
  }

  override def towardsWrap(headingX: Double, headingY: Double): Double = {
    val headingX2 = Topology.wrap(headingX,
      world.worldWidth / -2.0,
      world.worldWidth / 2.0)
    if (headingY == 0)
      if (headingX2 > 0) 90 else 270
    else if (headingX2 == 0)
      if (headingY > 0) 0 else 180
    else
      ((270 + StrictMath.toDegrees(StrictMath.PI + StrictMath.atan2(-headingY, headingX2)))
        % 360)
  }

  override def shortestPathX(x1: Double, x2: Double): Double = {
    val xprime =
      if (x1 > x2)
        x2 + world.worldWidth
      else
        x2 - world.worldWidth
    if (StrictMath.abs(x2 - x1) > StrictMath.abs(xprime - x1))
      xprime
    else
      x2
  }

  override def shortestPathY(y1: Double, y2: Double) = y2

  override def followOffsetY = 0.0

  override def getPN(source: Patch): Patch =
    if (source.pycor == world.maxPycor)
      null
    else
      world.fastGetPatchAt(source.pxcor, source.pycor + 1)
  override def getPE(source: Patch): Patch =
    world.fastGetPatchAt(
      if (source.pxcor == world.maxPxcor)
        world.minPxcor
      else
        source.pxcor + 1,
      source.pycor)
  override def getPS(source: Patch): Patch =
    if (source.pycor == world.minPycor)
      null
    else
      world.fastGetPatchAt(source.pxcor, source.pycor - 1)
  override def getPW(source: Patch): Patch =
    world.fastGetPatchAt(
      if (source.pxcor == world.minPxcor)
        world.maxPxcor
      else
        source.pxcor - 1,
      source.pycor)
  override def getPNE(source: Patch): Patch =
    if (source.pycor == world.maxPycor)
      null
    else
      world.fastGetPatchAt(
        if (source.pxcor == world.maxPxcor)
          world.minPxcor
        else
          source.pxcor + 1,
        source.pycor + 1)
  override def getPSE(source: Patch): Patch =
    if (source.pycor == world.minPycor)
      null
    else
      world.fastGetPatchAt(
        if (source.pxcor == world.maxPxcor)
          world.minPxcor
        else
          source.pxcor + 1,
        source.pycor - 1)
  override def getPSW(source: Patch): Patch =
    if (source.pycor == world.minPycor)
      null
    else
      world.fastGetPatchAt(
        if (source.pxcor == world.minPxcor)
          world.maxPxcor
        else
          source.pxcor - 1,
        source.pycor - 1)
  override def getPNW(source: Patch): Patch =
    if (source.pycor == world.maxPycor)
      null
    else
      world.fastGetPatchAt(
        if (source.pxcor == world.minPxcor)
          world.maxPxcor
        else
          source.pxcor - 1,
        source.pycor + 1)

  @throws(classOf[AgentException])
  @throws(classOf[PatchException])
  override def diffuse(amount: Double, vn: Int) {
    val xx = world.worldWidth
    val yy = world.worldHeight
    val xx2 = xx * 2
    val yy2 = yy * 2
    val scratch = world.getPatchScratch
    val scratch2 = Array.ofDim[Double](xx, yy)
    val minx = world.minPxcor
    val miny = world.minPycor
    var x, y = 0

    try while(y < yy) {
      x = 0
      while (x < xx ) {
        scratch(x)(y) =
          world.fastGetPatchAt(x + minx, y + miny)
            .getPatchVariable(vn)
            .asInstanceOf[java.lang.Double].doubleValue
        scratch2(x)(y) = 0
        x += 1
      }
      y += 1
    }
    catch { case _: ClassCastException =>
      throw new PatchException(
        world.fastGetPatchAt(wrapX(x).toInt, wrapY(y).toInt))
    }

    y = yy
    while(y < yy2) {
      x = xx
      while (x < xx2) {
        val diffuseVal = (scratch(x - xx)(y - yy) / 8) * amount
        if (y > yy && y < yy2 - 1) {
          scratch2(x - xx)(y - yy) += scratch(x - xx)(y - yy) - (8 * diffuseVal)
          scratch2((x - 1) % xx)((y - 1) % yy) += diffuseVal
          scratch2((x - 1) % xx)(y % yy) += diffuseVal
          scratch2((x - 1) % xx)((y + 1) % yy) += diffuseVal
          scratch2(x % xx)((y + 1) % yy) += diffuseVal
          scratch2(x % xx)((y - 1) % yy) += diffuseVal
          scratch2((x + 1) % xx)((y - 1) % yy) += diffuseVal
          scratch2((x + 1) % xx)(y % yy) += diffuseVal
          scratch2((x + 1) % xx)((y + 1) % yy) += diffuseVal
        } else if (y == yy) {
          scratch2(x - xx)(y - yy) += scratch(x - xx)(y - yy) - (5 * diffuseVal)
          scratch2((x - 1) % xx)(y % yy) += diffuseVal
          scratch2((x - 1) % xx)((y + 1) % yy) += diffuseVal
          scratch2(x % xx)((y + 1) % yy) += diffuseVal
          scratch2((x + 1) % xx)(y % yy) += diffuseVal
          scratch2((x + 1) % xx)((y + 1) % yy) += diffuseVal
        } else {
          scratch2(x - xx)(y - yy) += scratch(x - xx)(y - yy) - (5 * diffuseVal)
          scratch2((x - 1) % xx)(y % yy) += diffuseVal
          scratch2((x - 1) % xx)((y - 1) % yy) += diffuseVal
          scratch2(x % xx)((y - 1) % yy) += diffuseVal
          scratch2((x + 1) % xx)(y % yy) += diffuseVal
          scratch2((x + 1) % xx)((y - 1) % yy) += diffuseVal
        }
        x += 1
      }
      y += 1
    }
    y = 0
    while (y < yy) {
      x = 0
      while (x < xx) {
        if (scratch2(x)(y) != scratch(x)(y))
          world.getPatchAtWrap(x + minx, y + miny)
              .setPatchVariable(vn, Double.box(scratch2(x)(y)))
        x += 1
      }
      y += 1
    }
  }

  @throws(classOf[AgentException])
  @throws(classOf[PatchException])
  override def diffuse4(amount: Double, vn: Int) {
    val xx = world.worldWidth
    val yy = world.worldHeight
    val xx2 = xx * 2
    val yy2 = yy * 2
    val scratch = world.getPatchScratch
    val scratch2 = Array.ofDim[Double](xx, yy)
    val minx = world.minPxcor
    val miny = world.minPycor
    var x, y = 0
    try while (y < yy) {
      x = 0
      while (x < xx) {
        scratch(x)(y) =
          world.fastGetPatchAt(x + minx, y + miny)
            .getPatchVariable(vn)
            .asInstanceOf[java.lang.Double].doubleValue
        scratch2(x)(y) = 0
        x += 1
      }
      y += 1
    }
    catch { case _: ClassCastException =>
      throw new PatchException(
        world.fastGetPatchAt(wrapX(x).toInt, wrapY(y).toInt))
    }
    y = yy
    while (y < yy2) {
      x = xx
      while (x < xx2) {
        val diffuseVal = (scratch(x - xx)(y - yy) / 4) * amount
        if (y > yy && y < yy2 - 1) {
          scratch2(x - xx)(y - yy) += scratch(x - xx)(y - yy) - (4 * diffuseVal)
          scratch2((x - 1) % xx)(y % yy) += diffuseVal
          scratch2(x % xx)((y + 1) % yy) += diffuseVal
          scratch2(x % xx)((y - 1) % yy) += diffuseVal
          scratch2((x + 1) % xx)(y % yy) += diffuseVal
        } else if (y == yy) {
          scratch2(x - xx)(y - yy) += scratch(x - xx)(y - yy) - (3 * diffuseVal)
          scratch2((x - 1) % xx)(y % yy) += diffuseVal
          scratch2(x % xx)((y + 1) % yy) += diffuseVal
          scratch2((x + 1) % xx)(y % yy) += diffuseVal
        } else {
          scratch2(x - xx)(y - yy) += scratch(x - xx)(y - yy) - (3 * diffuseVal)
          scratch2((x - 1) % xx)(y % yy) += diffuseVal
          scratch2(x % xx)((y - 1) % yy) += diffuseVal
          scratch2((x + 1) % xx)(y % yy) += diffuseVal
        }
        x += 1
      }
      y += 1
    }
    y = 0
    while (y < yy) {
      x = 0
      while(x < xx) {
        if (scratch2(x)(y) != scratch(x)(y))
          world.getPatchAtWrap(x + minx, y + miny)
            .setPatchVariable(vn, Double.box(scratch2(x)(y)))
        x += 1
      }
      y += 1
    }
  }

}
